src/core/stream/video_events.ts

```diff
@@ -15,81 +15,123 @@
 import log from "../../log";
 
 /**
  * Set the initial time given as soon as possible on the video element.
  * Emit when done.
- * @param {HMTLMediaElement} videoElement
- * @param {number|Function} startTime
+ * @param {HMTLMediaElement} mediaElement
+ * @param {number|Function} startTime - Initial starting position. As seconds
+ * or as a function returning seconds.
  * @returns {Observable}
  */
 function doInitialSeek(
-  videoElement : HTMLMediaElement,
+  mediaElement : HTMLMediaElement,
   startTime : number|(() => number)
 ) : Observable<void> {
-  return hasLoadedMetadata(videoElement)
+  return hasLoadedMetadata(mediaElement)
     .pipe(
       tap(() => {
         log.info("set initial time", startTime);
 
         // reset playbackRate to 1 in case we were at 0 (from a stalled
         // retry for instance)
-        videoElement.playbackRate = 1;
-        videoElement.currentTime = typeof startTime === "function" ?
+        mediaElement.playbackRate = 1;
+        mediaElement.currentTime = typeof startTime === "function" ?
           startTime() : startTime;
       }),
       shareReplay() // we don't want to repeat the side-effect on each
                     // subscription of this very same observable
     );
 }
 
+interface ISeekedEvent {
+  type : "seeked";
+  value : undefined;
+}
+
+interface IPlayedEvent {
+  type : "played";
+  value : "done"|"blocked";
+}
+
+interface ILoadedEvent {
+  type : "loaded";
+  value : undefined;
+}
+
+export type IMediaReadyEvent =
+  ISeekedEvent |
+  IPlayedEvent |
+  ILoadedEvent;
+
 /**
- * @param {HTMLMediaElement} videoElement
- * @param {number|Function} startTime
- * @param {boolean} autoPlay
+ * @param {HTMLMediaElement} mediaElement
+ * @param {number|Function} startTime - Initial starting position. As seconds
+ * or as a function returning seconds.
+ * @param {boolean} autoPlay - Whether the player should auto-play
  * @returns {object}
  */
-export default function handleVideoEvents(
-  videoElement : HTMLMediaElement,
+export default function seekAndPlayOnMediaReady(
+  mediaElement : HTMLMediaElement,
   startTime : number|(() => number),
   mustAutoPlay : boolean
-) : {
-  initialSeek$ : Observable<void>;
-  loadAndPlay$ : Observable<void>;
-} {
-  const initialSeek$ = doInitialSeek(videoElement, startTime);
-  const handledCanPlay$ = canPlay(videoElement).pipe(
-    tap(() => log.info("canplay event"))
-  );
+) : Observable<IMediaReadyEvent> {
+  const seek$ = doInitialSeek(mediaElement, startTime)
+    .pipe(mapTo({ type: "seeked" as "seeked", value : undefined }));
+
+  const loadAndPlay$ = canPlay(mediaElement).pipe(
+
+    tap(() => log.info("canplay event")),
+
+    mergeMap(() => {
+      const loadedEvent$ = observableOf({
+        type: "loaded" as "loaded",
+        value: undefined,
+      });
+
+      if (mustAutoPlay) {
+        const autoplay$ = play$(mediaElement).pipe(
+          mapTo({ type: "played" as "played", value: "done" as "done" }),
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=279&issues=AWUaMD4kopYCBC6oecGq&open=AWUaMD4kopYCBC6oecGq)

[ISSUE_KEY]: AWUaMD4kopYCBC6oecGq